bundles/org.openhab.binding.teleinfo/src/main/java/org/openhab/binding/teleinfo/internal/reader/io/TeleinfoInputStream.java

```diff
@@ -0,0 +1,682 @@
+        this(teleinfoInputStream, DEFAULT_TIMEOUT_WAIT_NEXT_HEADER_FRAME, DEFAULT_TIMEOUT_READING_FRAME,
+                autoRepairInvalidADPSgroupLine);
+    }
+
+    public TeleinfoInputStream(final InputStream teleinfoInputStream, long waitNextHeaderFrameTimeoutInMs,
+            long readingFrameTimeoutInMs, boolean autoRepairInvalidADPSgroupLine) {
+        if (teleinfoInputStream == null) {
+            throw new IllegalArgumentException("Teleinfo inputStream not null");
+        }
+
+        this.waitNextHeaderFrameTimeoutInMs = waitNextHeaderFrameTimeoutInMs;
+        this.readingFrameTimeoutInMs = readingFrameTimeoutInMs;
+        this.autoRepairInvalidADPSgroupLine = autoRepairInvalidADPSgroupLine;
+
+        try {
+            this.bufferedReader = new BufferedReader(new InputStreamReader(teleinfoInputStream, "ASCII"));
+            this.teleinfoInputStream = teleinfoInputStream;
+        } catch (UnsupportedEncodingException e) {
+            throw new IllegalStateException(e);
+        }
+
+        groupLine = null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        logger.debug("close() [start]");
+        bufferedReader.close();
+        executorService.shutdownNow();
+        super.close();
+        logger.debug("close() [end]");
+    }
+
+    /**
+     * Returns the next frame.
+     *
+     * @return the next frame or null if end of stream
+     * @throws InvalidFrameException if the read data from
+     * @throws TimeoutException if the delay to read a complete frame is expired (33,4 ms) or if the delay to find
+     *             the
+     *             header of next frame is expired (33,4 ms)
+     * @throws IOException
+     */
+    public synchronized Frame readNextFrame() throws InvalidFrameException, TimeoutException, IOException {
+        logger.debug("readNextFrame() [start]");
+
+        // seek the next header frame
+        Future<Void> seekNextHeaderFrameTask = executorService.submit(new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+                while (!isHeaderFrame(groupLine)) {
+                    groupLine = bufferedReader.readLine();
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("groupLine = {}", groupLine);
+                    }
+                    if (groupLine == null) { // end of stream
+                        logger.trace("end of stream reached !");
+                        return null;
+                    }
+                }
+
+                logger.trace("header frame found !");
+                return null;
+            }
+        });
+        try {
+            logger.debug("seeking the next header frame...");
+            logger.trace("waitNextHeaderFrameTimeoutInMs = {}", waitNextHeaderFrameTimeoutInMs);
+            seekNextHeaderFrameTask.get(waitNextHeaderFrameTimeoutInMs, TimeUnit.MICROSECONDS);
```

Comment:
The Teleinfo protocol defines a timeout between two frames

Reply:
If I see correctly the timeout is 33.4ms and you multiplied it with 100. So, the code in the `Callable` is interrupted after 3.3sec.. But I think the timeout never kicks in, since the timeout of the serial port is set to 250ms in TeleinfoSerialControllerHandler:177. So, you could remove the `sumbit()`s and execute the code directly.

Reply:
The Teleinfo protocol defines this timeout to 33.4ms, but with this default timeout, the binding throws always timeout errors. That's why, I have override the timeout to 3,34 s.

I had imagined setting this timeout to be configurable so that the user can overload the official timeout (33,4ms), but I am afraid that a new user does not think of modifying this parameter if errors occurred and that the user experience is very bad.

The timeout of the serial port (250ms) and this timeout are different. You could receive data between end of frame and start of frame due to, for example, hardware issues. In this case, the timeout of serial port will never be triggered when the delay could be triggered

Reply:
I agree with you that 33ms is too short for the JVM to handle it correctly and that the timeout shouldn't be configureable by the user. I guess you mean loose contact or so, that the serial port detects data continuously, where is no data. This case should be very rare, since you have a parity bit, so garbage will be dropped at operating system driver level. Is a correct communication possible at all, in this case?